nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/PutElasticsearchRecordTest.groovy

```diff
@@ -0,0 +1,272 @@
+
+        evalClosure = { List<IndexOperationRequest> items ->
+            def testTypeCount = items.findAll { it.type == "test_type" }.size()
+            def messageTypeCount = items.findAll { it.type == "message" }.size()
+            def testIndexCount = items.findAll { it.index == "test_index" }.size()
+            def bulkIndexCount = items.findAll { it.index.startsWith("bulk_") }.size()
+            def indexOperationCount = items.findAll { it.operation == IndexOperationRequest.Operation.Index }.size()
+            Assert.assertEquals(5, testTypeCount)
+            Assert.assertEquals(1, messageTypeCount)
+            Assert.assertEquals(5, testIndexCount)
+            Assert.assertEquals(1, bulkIndexCount)
+            Assert.assertEquals(6, indexOperationCount)
+        }
+
+        clientService.evalClosure = evalClosure
+        runner.enqueue(flowFileContents, [
+            "schema.name": "recordPathTest"
+        ])
+        runner.run()
+
+        runner.clearTransferState()
+
+        runner.setProperty(PutElasticsearchRecord.OPERATION_RECORD_PATH, "/operation")
+
+        flowFileContents = prettyPrint(toJson([
+            [ id: "rec-1", index: "bulk_a", type: "message", msg: "Hello", operation: "index" ],
+            [ id: "rec-2", index: "bulk_b", type: "message", msg: "Hello", operation: "Index" ],
+            [ id: "rec-3", index: "bulk_a", type: "message", msg: "Hello", operation: "DeLEte" ],
+            [ id: "rec-4", index: "bulk_b", type: "message", msg: "Hello", operation: "updATE" ],
+            [ id: "rec-5", index: "bulk_a", type: "message", msg: "Hello", operation: "upsert" ],
+            [ id: "rec-6", index: "bulk_b", type: "message", msg: "Hello", operation: "uPSERT" ]
+        ]))
+
+        clientService.evalClosure = { List<IndexOperationRequest> items ->
+            int index = items.findAll { it.operation == IndexOperationRequest.Operation.Index }.size()
+            int delete = items.findAll { it.operation == IndexOperationRequest.Operation.Delete }.size()
+            int update = items.findAll { it.operation == IndexOperationRequest.Operation.Update }.size()
+            int upsert = items.findAll { it.operation == IndexOperationRequest.Operation.Upsert }.size()
+            Assert.assertEquals(2, index)
+            Assert.assertEquals(1, delete)
+            Assert.assertEquals(1, update)
+            Assert.assertEquals(2, upsert)
+        }
+
+        runner.enqueue(flowFileContents, [
+            "schema.name": "recordPathTest"
+        ])
+        runner.run()
+        runner.assertTransferCount(PutElasticsearchRecord.REL_SUCCESS, 1)
+        runner.assertTransferCount(PutElasticsearchRecord.REL_FAILURE, 0)
+        runner.assertTransferCount(PutElasticsearchRecord.REL_RETRY, 0)
+    }
+
+    @Test
+    void testInputRequired() {
+        runner.run()
+        runner.assertTransferCount(PutElasticsearchRecord.REL_SUCCESS, 0)
+    }
+
+    @Test
+    void testErrorRelationship() {
```

Comment:
Might be nice to test the non-happy path here, where the error record writer has either a bad and/or different schema (rather than using the same schema which is technically like inheriting it). Like what if it had an extra non-nullable field, does everything still work as expected (rollback vs. failure, e.g.)?

Reply:
When is that likely to happen?